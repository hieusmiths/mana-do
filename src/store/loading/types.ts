export const AUTH_SIGNIN_REQUEST = 'auth_signin_request';
export const AUTH_SIGN_IN_FAILURE = 'auth_signin-failure';