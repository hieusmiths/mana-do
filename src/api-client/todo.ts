const BASE: string = '/todos';

export const API_TODO_BASE = BASE;