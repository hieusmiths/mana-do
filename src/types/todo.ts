export enum ETodoStatus {
  ACTIVE = 'ACTIVE',
  COMPLETED = 'COMPLETED'
}

export interface ITodo {
  id: string
  user_id: string
  content : string
  status?: ETodoStatus
  created_date: string
}