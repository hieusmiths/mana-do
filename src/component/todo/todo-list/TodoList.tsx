import React from 'react';
import TodoItem from '../todo-item';
import { ITodo } from '../../../types/todo';
import { isTodoCompleted } from '../../../helper';

interface Props {
  list: ITodo[],
  onDelete: Function,
  onUpdateTodoStatus: Function,
  onUpdate: (todoId: string, content: string) => void
}

const TodoList = ({ list, onDelete, onUpdateTodoStatus, onUpdate }: Props) => {
  if (!list.length) {
    return <div>Your Jobs is empty!</div>
  }
  return (
    <div className="ToDo__list">
    {
      list.map((item: ITodo, index) => {
        const { id, content } = item;
        const isDone = isTodoCompleted(item);
        return <TodoItem key={index} id={id}
          content={content}
          isDone={isDone}
          onDelete={() => onDelete(id)}
          onUpdateStatus={onUpdateTodoStatus}
          onUpdate={onUpdate}
          /> 
      })
    }
    </div>)
};

export default TodoList;