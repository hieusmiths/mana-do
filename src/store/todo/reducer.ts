import { ITodo, ETodoStatus } from '../../types/todo';
import {
  AppActions,
  CREATE_TODO,
  DELETE_ALL_TODOS,
  DELETE_TODO,
  TOGGLE_ALL_TODOS,
  UPDATE_TODO_STATUS,
  UPDATE_TODO,
  SET_TODO
} from './actions';

export interface AppState {
  todos: ITodo[]
}

export const initialState: AppState = {
  todos: []
}

function reducer(state: AppState, action: AppActions): AppState {
  switch (action.type) {
    case SET_TODO: {
      return { ...state, todos: action.payload };
    }
    case CREATE_TODO:
      return {
        ...state,
        todos: [...state.todos, action.payload]
      };

    case UPDATE_TODO_STATUS: {
      const { todoId, checked } = action.payload;
      const updatedIndex = state.todos.findIndex((todo) => todo.id === todoId);
      const updatedTodo = { ...state.todos[updatedIndex], status: checked ? ETodoStatus.COMPLETED : ETodoStatus.ACTIVE }
      return {
        ...state,
        todos: [
          ...state.todos.slice(0, updatedIndex), updatedTodo, ...state.todos.slice(updatedIndex + 1)
        ]
      }
    }
    case TOGGLE_ALL_TODOS:
      const tempTodos = state.todos.map((e)=>{
        return {
          ...e,
          status: action.payload ? ETodoStatus.COMPLETED : ETodoStatus.ACTIVE
        }
      })

      return {
        ...state,
        todos: tempTodos
      }

    case DELETE_TODO:
      const deletedIndex = state.todos.findIndex((todo) => todo.id === action.payload);
      const remainTodos = [...state.todos.slice(0, deletedIndex), ...state.todos.slice(deletedIndex + 1)];
      return {
        ...state,
        todos: remainTodos
      }
    case DELETE_ALL_TODOS:
      return {
        ...state,
        todos: []
      }
    case UPDATE_TODO: {
      const updatedIndex = state.todos.findIndex((todo) => todo.id === action.payload.todoId);
      const updatedTodos = [...state.todos.slice(0, updatedIndex), action.payload.updatedTodo, ...state.todos.slice(updatedIndex + 1)]
      return {
        ...state,
        todos: updatedTodos
      }
    }
    default:
      return state;
  }
}

export default reducer;