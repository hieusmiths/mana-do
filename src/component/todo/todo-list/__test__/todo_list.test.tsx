import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TodoList, { Props } from '../TodoList';
import { ITodo, ETodoStatus } from '../../../../types/todo';

describe('<TodoList />', () => {
  const onDeteleMock = jest.fn();
  const onUpdateMock = jest.fn();
  const onUpdateStatusMock = jest.fn();
  const listMock: ITodo[] = [
    {
      id: '1',
      content: 'one',
      status: ETodoStatus.ACTIVE,
      created_date: new Date().toISOString(),
      user_id: 'user'
    },
  ];
  const defaultProps: Props = {
    list: listMock,
    onDelete: onDeteleMock,
    onUpdate: onUpdateMock,
    onUpdateTodoStatus: onUpdateStatusMock
  }
  it('Should render corectly', () => {
    const wrapper = shallow(<TodoList {...defaultProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render empty', () => {
    const wrapper = shallow(<TodoList {...defaultProps} list={[]} />);
    expect(wrapper.find('.empty').exists()).toBeTruthy();
  });

  it('Should render n TodoItem', () => {
    const wrapper = shallow(<TodoList {...defaultProps} />);
    expect(wrapper.find('TodoItem')).toHaveLength(1);
  });

});