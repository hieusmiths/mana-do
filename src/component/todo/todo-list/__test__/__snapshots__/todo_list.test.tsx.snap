// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> Should render corectly 1`] = `
<div
  className="ToDo__list"
>
  <TodoItem
    content="one"
    id="1"
    isDone={false}
    key="0"
    onDelete={[Function]}
    onUpdate={[MockFunction]}
    onUpdateStatus={[MockFunction]}
  />
</div>
`;
